### PR TITLE
Disable protocol and TLS hints on skipped ports

### LINKED ISSUE #5995: Proxy tries to initiate TLS on skipped inbound port application
While investigating #5913, it looks like the outbound side of the proxy tries to proxy a TLS connection on a server that has its `skip-inbound-ports` configured. It seems as if the outbound proxy gets discovery info about where to send the request from the `destination` service, however, this info does not  include info about the server's `skip-inbound-ports` configuration.

The outbound proxy then tries to establish a TLS connection and then prints this error message:
```
linkerd-proxy [   611.222635s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:34902}: linkerd_app_core::serve: Connection closed reason=received corrupt message    
```

Here is a full log showing this behavior:
```
[    30.806144s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:36802}:server{orig_dst=10.43.166.126:4444}:poll_profile:h2:Connection{peer=Client}: h2::codec::framed_write: send frame=Headers { stream_id: StreamId(5), flags: (0x4: END_HEADERS) }
[    30.806170s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:36802}:server{orig_dst=10.43.166.126:4444}:poll_profile:h2:Connection{peer=Client}: h2::codec::framed_write: send frame=Data { stream_id: StreamId(5) }
[    30.806176s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:36802}:server{orig_dst=10.43.166.126:4444}:poll_profile:h2:Connection{peer=Client}: h2::codec::framed_write: send frame=Data { stream_id: StreamId(5), flags: (0x1: END_STREAM) }
[    30.806260s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:36802}:server{orig_dst=10.43.166.126:4444}:poll_profile: trust_dns_proto::udp::udp_stream: created socket successfully
[    30.807188s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:36802}:server{orig_dst=10.43.166.126:4444}:poll_profile: trust_dns_proto::udp::udp_client_stream: received message id: 20600
[    30.807239s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:36802}:server{orig_dst=10.43.166.126:4444}:poll_profile: linkerd_dns: addrs=[10.42.0.10:8086]
[    30.807247s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:36802}:server{orig_dst=10.43.166.126:4444}:poll_profile: linkerd_proxy_dns_resolve: addrs=[10.42.0.10:8086]
[    30.807351s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:36802}:server{orig_dst=10.43.166.126:4444}:poll_profile:h2:Connection{peer=Client}: h2::codec::framed_read: received frame=Headers { stream_id: StreamId(5), flags: (0x4: END_HEADERS) }
[    30.807428s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:36802}:server{orig_dst=10.43.166.126:4444}:poll_profile:h2:Connection{peer=Client}: h2::codec::framed_read: received frame=Data { stream_id: StreamId(5) }
[    30.807487s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:44290}:server{orig_dst=10.43.166.126:4444}:tcp: tower::balance::p2c::service: updating from discover
[    30.807545s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:44290}:server{orig_dst=10.43.166.126:4444}:tcp: linkerd_proxy_api_resolve::resolve: Add endpoints=1
[    30.807587s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:44290}:server{orig_dst=10.43.166.126:4444}:tcp: tower::balance::p2c::service: updating from discover
[    30.807623s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:44290}:server{orig_dst=10.43.166.126:4444}:tcp: tower::buffer::worker: service.ready=true processing request
[    30.807652s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:44290}:server{orig_dst=10.43.166.126:4444}:tcp:endpoint{server.addr=10.42.0.170:4444 server.id=ServerId("default.default.serviceaccount.identity.linkerd.cluster.local")}: linkerd_tls::client: Initiating TLS connection server.id=default.default.serviceaccount.identity.linkerd.cluster.local
[    30.807663s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:44290}:server{orig_dst=10.43.166.126:4444}:tcp:endpoint{server.addr=10.42.0.170:4444 server.id=ServerId("default.default.serviceaccount.identity.linkerd.cluster.local")}: linkerd_proxy_transport::connect: Connecting server.addr=10.42.0.170:4444
[    30.807918s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:44290}:server{orig_dst=10.43.166.126:4444}:tcp:endpoint{server.addr=10.42.0.170:4444 server.id=ServerId("default.default.serviceaccount.identity.linkerd.cluster.local")}: linkerd_proxy_transport::connect: Connected local.addr=10.42.0.169:42828 keepalive=Some(10s)
[    30.807946s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:44290}:server{orig_dst=10.43.166.126:4444}:tcp:endpoint{server.addr=10.42.0.170:4444 server.id=ServerId("default.default.serviceaccount.identity.linkerd.cluster.local")}: rustls::client::hs: No cached session for DNSNameRef("default.default.serviceaccount.identity.linkerd.cluster.local")
[    30.807952s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:44290}:server{orig_dst=10.43.166.126:4444}:tcp:endpoint{server.addr=10.42.0.170:4444 server.id=ServerId("default.default.serviceaccount.identity.linkerd.cluster.local")}: rustls::client::hs: Not resuming any session
[    30.808167s] DEBUG ThreadId(01) outbound:accept{client.addr=10.42.0.169:44290}: linkerd_app_core::serve: Connection closed reason=received corrupt message
[    30.808180s] DEBUG ThreadId(01) evict{target=Logical { orig_dst: OrigDstAddr(10.43.166.126:4444), protocol: (), profile: Some(..) }}: linkerd_cache: Awaiting idleness
```

### How to reproduce
- `git clone git@github.com:olix0r/tcp-echo.git`
- `helm package .`
- `helm install tcp-echo tcp-echo-0.1.0.tgz`
- Annotate the pod template spec of the client and server deployments with `config.linkerd.io/proxy-log-level: debug`
- Annotate only the server's pod template spec with `config.linkerd.io/skip-inbound-ports: 4444`
- `kubectl rollout restart deploy`

### What to expect
Ideally, the outbound proxy shouldn't proxy the connection with mTLS if the server has a skipped port. We would probably need to modify destination to send that info and then modify the proxy to detect that and disable mTLS only for that host:port connection.

### PR DESCRIPTION
**Pull request by dadjeibaah**
When a pod is configured with `skip-inbound-ports` annotation, a client
proxy trying to connect to that pod tries to connect to it via H2 and
also tries to initiate a TLS connection. This issue is caused by the
destination controller when it sends protocol and TLS hints to the
client proxy for that skipped port.

This change fixes the destination controller so that it no longer
sends protocol and TLS identity hints to outbound proxies resolving a
`podIP:port` that is on a skipped inbound port.

I've included a test that exhibits this error prior to this fix but you
can also test the prior behavior by:

```bash
curl https://run.linkerd.io/booksapp.yml > booksapp.yaml

# edit either the books or authors service to:
1: Configure a failure rate of 0.0
2: add the `skip-inbound-ports` config annotation

bin/linkerd viz stat pods webapp

There should be no successful requests on the webapp deployment
```
Fixes #5995

### CODE REVIEW

File: controller/api/destination/server.go [hunk 1]
```diff
@@ -231,7 +231,13 @@ func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetPr
 				if !ok {
 					opaquePorts = s.defaultOpaquePorts
 				}
-				endpoint, err = toWeightedAddr(podSet.Addresses[podID], opaquePorts, s.enableH2Upgrade, s.identityTrustDomain, s.controllerNS, log)
+
+				skippedInboundPorts, err := getPodSkippedInboundPortsAnnotations(pod)
+				if err != nil {
+					log.Errorf("failed getting ignored inbound ports annoatation for pod: %s", err)
```
Comment (kleimkuhler) [inline controller/api/destination/server.go:237]: Typo, and then match the way we word other errors in this function:
```suggestion
					log.Errorf("failed to get ignored inbound ports annotation for pod: %s", err)
```

File: controller/api/destination/server.go [hunk 2]
```diff
@@ -482,3 +488,16 @@ func getPodOpaquePortsAnnotations(pod *corev1.Pod) (map[uint32]struct{}, bool, e
...
+	if !ok {
+		return nil, nil
+	}
+
+	if annotation == "" {
+		return map[uint32]struct{}{}, nil
+	}
+
+	return util.ParsePorts(annotation)
+}
```
Comment (kleimkuhler) [inline controller/api/destination/server.go:503]: `toWeightedAddr` checks if this return value contains the `port` key. The map does not need to be initialized for this check, so it can return `nil` when `annotation == ""` as well:
```suggestion
func getPodSkippedInboundPortsAnnotations(pod *corev1.Pod) (map[uint32]struct{}, error) {
	annotation, ok := pod.Annotations[labels.ProxyIgnoreInboundPortsAnnotation]
	if !ok || annotation == "" {
		return nil, nil
	}
	return util.ParsePorts(annotation)
}
```

**Review by kleimkuhler — COMMENTED** [review]
Overall this looks good. I have a few smaller comments.

The integration test looks good, but I'd still like to see a unit test in `server_test.go` that actually looks at the destination response and asserts that `Endpoint.ProtocolHint == nil` and `Endpoint.Identity == nil`. These should more directly test this change.

File: test/integration/skipports/skip_ports_test.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+//////////////////////
+/// TEST EXECUTION ///
+//////////////////////
+
+func TestSkipInboundPorts(t *testing.T) {
+	ctx := context.Background()
+
+	err := TestHelper.CreateDataPlaneNamespaceIfNotExists(ctx, skipPortsNs, nil)
```
Comment (alpeb) [inline test/integration/skipports/skip_ports_test.go:36]: Can I suggest to use `TestHelper.WithDataPlaneNamespace()` instead? As we keep adding integration tests, when they are run sequentially (like in the ARM environment), the cluster sometimes runs out of resources. `WithDataPlaneNamespace()` will clean up resources when the test is done.

File: test/integration/skipports/skip_ports_test.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+		cmd := []string{"diagnostics", "proxy-metrics", "--namespace", skipPortsNs, podName}
+
+		metrics, err := TestHelper.LinkerdRun(cmd...)
+		if err != nil {
+			testutil.AnnotatedFatalf(t, "error getting metrics for pod", "error getting metrics for pod\n%s", err)
+		}
+
+		if httpResponseTotalMetricRE.MatchString(metrics) {
+			testutil.AnnotatedFatalf(t, "expected not to find HTTP outbound response failures to dst=books.skip-ports-test.svc.cluster.local:7002",
+				"expected not to find HTTP outbound requests when pod is opaque\n%s", metrics)
```
Comment (alpeb) [inline test/integration/skipports/skip_ports_test.go:82]:
```suggestion
				"expected not to find HTTP outbound requests when pod is skipping inbound ports\n%s", metrics)
```

**Review by alpeb — COMMENTED** [review]
Looking good, modulo @klingerf's suggestions :+1:

File: controller/api/destination/server_test.go [hunk 5]
```diff
@@ -760,6 +850,44 @@ func TestGetProfiles(t *testing.T) {
 			t.Fatalf("Expected port %d to be an opaque protocol, but it was not", opaquePort)
 		}
 	})
+
+	t.Run("Return profile with unknown protocol hint and identity when service name contains skipped inbound port", func(t *testing.T) {
```
Comment (kleimkuhler) [inline controller/api/destination/server_test.go:854]: Getting profile for pod IP here, so maybe reword:
```suggestion
	t.Run("Return profile with unknown protocol hint and identity when pod contains skipped inbound port", func(t *testing.T) {
```

**Review by kleimkuhler — APPROVED** [review]
Looks good @dadjeibaah thanks!

**Review by alpeb — APPROVED** [review]
Thanks @dadjeibaah , LGTM! :+1:

**Review by adleong — APPROVED** [review]
